src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clear(stateCopy);
+        break;
+
+      default:
+        return 'Error';
+    }
+
+    result.push({ ...stateCopy });
+  }
+
+  return result;
+}
+
+function addProperties(state, extraData) {
+  Object.assign(state, extraData);
+}
+
+function removeProperties(state, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete state[key];
+  }
+}
```

Comment:
[CODE QUALITY]: Similar to the `addProperties` function, `removeProperties` also directly modifies the passed `state` object. While it works for this task, be cautious about directly mutating function inputs in general.